clang/lib/CodeGen/CGHLSLRuntime.cpp

```diff
@@ -446,6 +449,13 @@ void CGHLSLRuntime::setHLSLFunctionAttributes(const FunctionDecl *FD,
     const StringRef ExportAttrKindStr = "hlsl.export";
     Fn->addFnAttr(ExportAttrKindStr);
   }
+  llvm::Triple T(Fn->getParent()->getTargetTriple());
+  if (T.getEnvironment() == llvm::Triple::EnvironmentType::Library) {
+    if (CGM.getCodeGenOpts().OptimizationLevel == 0) {
+      Fn->addFnAttr(llvm::Attribute::OptimizeNone);
+      Fn->addFnAttr(llvm::Attribute::NoInline);
+    }
+  }
```

Comment:
Do we want to do this on all functions in a library or just entry points and exported functions? In any case, it really would be preferable if "SetLLVMFunctionAttributesForDefinition" did the right thing (whatever that may be) rather than us needing to duplicate that logic here...